File: 19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        lenl = 1
        tmp0 = head
        while tmp0 != None:
            print tmp0.val, lenl
            tmp0 = tmp0.next
            lenl = lenl + 1
        
        print("================")
        i = 1
        tmp1 = head
        while (lenl - n) - i > 0:
            print i, tmp1.val
            i = i + 1
            tmp1 = tmp1.next
        print tmp1.val
        print("================")

File: 19. Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *fp = head, *sp = head; //use two pointers
        for (int i = 0; i < n; i++)
            fp = fp -> next;
        if (fp == NULL)
            return head -> next;
        while (fp -> next != NULL){
            fp = fp -> next;
            sp = sp -> next;
        }
        sp -> next = sp -> next -> next;
        return head;
    }
};
